lucene/core/src/java/org/apache/lucene/search/comparators/DocComparator.java

```diff
@@ -0,0 +1,183 @@
+        private final int docBase;
+        private final int minDoc;
+        private final int maxDoc;
+        private DocIdSetIterator competitiveIterator; // iterator that starts from topValue
+
+        public DocLeafComparator(LeafReaderContext context) {
+            this.docBase = context.docBase;
+            if (enableSkipping) {
+                this.minDoc = topValue + 1;
+                this.maxDoc = context.reader().maxDoc();
+                this.competitiveIterator = DocIdSetIterator.all(maxDoc);
+            } else {
+                this.minDoc = -1;
+                this.maxDoc = -1;
+                this.competitiveIterator = null;
+            }
+        }
+
+        @Override
+        public void setBottom(int slot) {
+            bottom = docIDs[slot];
+            bottomValueSet = true;
+            updateIterator();
+        }
+
+        @Override
+        public int compareBottom(int doc) {
+            // No overflow risk because docIDs are non-negative
+            return bottom - (docBase + doc);
+        }
+
+        @Override
+        public int compareTop(int doc) {
+            int docValue = docBase + doc;
+            return Integer.compare(topValue, docValue);
+        }
+
+        @Override
+        public void copy(int slot, int doc) throws IOException {
+            docIDs[slot] = docBase + doc;
+        }
+
+        @Override
+        public void setScorer(Scorable scorer) throws IOException {
+            // update an iterator on a new segment
+            updateIterator();
+        }
+
+        @Override
+        public DocIdSetIterator competitiveIterator() {
+            if (enableSkipping == false) {
+                return null;
+            } else {
+                return new DocIdSetIterator() {
+                    private int doc;
+
+                    @Override
+                    public int nextDoc() throws IOException {
+                        return doc = competitiveIterator.nextDoc();
+                    }
+
+                    @Override
+                    public int docID() {
+                        return doc;
+                    }
+
+                    @Override
+                    public long cost() {
+                        return competitiveIterator.cost();
+                    }
+
+                    @Override
+                    public int advance(int target) throws IOException {
+                        return doc = competitiveIterator.advance(target);
+                    }
+                };
+            }
+        }
+
+        @Override
+        public void setHitsThresholdReached() {
+            hitsThresholdReached = true;
+            updateIterator();
+        }
+
+        private void updateIterator() {
+            if (enableSkipping == false || hitsThresholdReached == false) return;
+            if (bottomValueSet) {
+                // since we've collected top N matches, we can early terminate
```

Comment:
Can you add a comment explaining that early termination is already implemented in the collector but we'll remove in a follow up ?

Reply:
addressed in 485fe4f